forest_benchmarking/classical_logic/ripple_carry_adder.py

```diff
@@ -0,0 +1,367 @@
+"""
+Circuit primitives for classical reversible logic
+
+At the moment it is primarily using the simple adder construction in
+
+[CDKM96] A new quantum ripple-carry addition circuit
+         Cuccaro, Draper, Kutin, and Moulton
+         https://arxiv.org/abs/quant-ph/0410184
+
+There are many other classical logic primitives that can be coded see
+e.g.
+
+[VBE96] Quantum networks for elementary arithmetic operations
+        Vedral,  Barenco, Ekert
+        Phys. Rev. A 54, 147 (1996)
+        https://doi.org/10.1103/PhysRevA.54.147
+        https://arxiv.org/abs/quant-ph/9511018
+"""
+from typing import Sequence, Tuple
+import networkx as nx
+import numpy as np
+from numpy import pi
+from scipy.spatial.distance import hamming
+
+from pyquil.gates import CNOT, CCNOT, X, I, H, CZ, MEASURE, RESET
+from pyquil import Program
+from pyquil.api import QuantumComputer
+from pyquil.unitary_tools import all_bitstrings
+
+from forest_benchmarking.readout import _readout_group_parameterized_bitstring
+from forest_benchmarking.classical_logic.primitives import *
+
+
+def assign_registers_to_line_or_cycle(start: int, graph: nx.Graph, num_length: int):
```

Comment:
I was hoping to do something smarter but I couldn't find a nice way to do it, even assuming that I was trying to map a line onto tiled octagons/squares (or really even if I assumed Aspen topology). I was attempting a deterministic approach; however, I have some code that samples connected sub-graphs of a given size at random using Metropolis-Hastings which I could try to modify to prefer arrangements with smaller distances between qubits adjacent in the circuit. Let me know if this seems worth adding. Right now you can always eye-ball a line and remove nodes from the circuit topology until you get close enough without too much trouble.